### PR TITLE
Single Model Pseudo Labeling

### PR DESCRIPTION
**Pull request by DolanTheMFWizard**
*Issue #, if available:*

*Description of changes:*
Added in pseudo labeling algorithm for single model, bagged modeling connection still in progress.

NOTE: This is the basic pseudo labeling without calibration adjustment or the multi-model regression pseudo labeling with better performance over random selection.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1322-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/2/index.html

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 1]
```diff
@@ -849,13 +853,15 @@ def fit_extra(self, hyperparameters, time_limit=None, base_model_names=None, **k
         ag_args_fit = kwargs['ag_args_fit']
         ag_args_ensemble = kwargs['ag_args_ensemble']
         excluded_model_types = kwargs['excluded_model_types']
+        X_pseudo = kwargs.get('X_pseudo', None)
+        y_pseudo = kwargs.get('y_pseudo', None)
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:857]: shouldn't this just be a dataframe of the same format as `train_data` argument? what is the advantage of separating it into X and y (which are not arguments of `predictor` normally)

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 2]
```diff
@@ -898,6 +909,98 @@ def fit_extra(self, hyperparameters, time_limit=None, base_model_names=None, **k
...
+            if previous_score >= current_score:
+                return self
+            else:
+                # Cut down X_test to not include pseudo labeled data
+                X_test = X_test.loc[test_pseudo_idxes[test_pseudo_idxes == False].index]
+                previous_score = current_score
+
+        return self
+
+    def pseudo_label_fit(self, test_data: pd.DataFrame, max_iter: int = 5, **kwargs):
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:978]: for future extensions to semi-supervised learning, shouldn't `test_data` be renamed to something more general like: `pseudolabeled_data`?

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 3]
```diff
@@ -898,6 +909,98 @@ def fit_extra(self, hyperparameters, time_limit=None, base_model_names=None, **k
         self.save()
         return self
 
+
+    def _filter_pseudo(self, y_pred_proba_og, min_percentage: float = 0.05, max_percentage: float = 0.6,
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:913]: will be good to have 1 line of documentation describing what each method is intended to do.

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 4]
```diff
@@ -2342,6 +2451,38 @@ def _validate_fit_extra_kwargs(self, kwargs, extra_valid_keys=None):
...
+            if self.sample_weight in train_features:
+                train_features.remove(self.sample_weight)
+            if self.sample_weight in other_features:
+                other_features.remove(self.sample_weight)
+        if self._learner.groups is not None and is_labeled:
+            train_features.remove(self._learner.groups)
+
+        return train_features, other_features
+
+    def _validate_pseudo_data(self, train_data: pd.DataFrame, pseudo_data: pd.DataFrame):
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:2465]: I find "pseudolabeled" to be much clearer than the shorter prefix "pseudo". 
"pseudo_data" seems quite vague to me as a variable name compared to "pseudolabeled_data"

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 5]
```diff
@@ -898,6 +909,98 @@ def fit_extra(self, hyperparameters, time_limit=None, base_model_names=None, **k
...
+        if isinstance(hyperparameters, str):
+            hyperparameters = get_hyperparameter_config(hyperparameters)
+
+        if is_labeled:
+            X_pseudo, y_pseudo, _, _, _, _, _, _ = self._learner.general_data_processing(test_data, None, None, 1, 0)
+            kwargs['X_pseudo'] = X_pseudo
+            kwargs['y_pseudo'] = y_pseudo
+            return self.fit_extra(hyperparameters=hyperparameters, name_suffix=PSEUDO_MODEL_SUFFIX.format(iter=1), **kwargs)
+        else:
+            return self._run_pseudo_labeling(hyperparameters=hyperparameters, test_data=test_data,
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:1001]: why is `run_pseudolabeling` a private method? Shouldn't the user be able to use this method after they've initially called `predictor.fit()`?

Comment (jwmueller) [reply tabular/src/autogluon/tabular/predictor/predictor.py:1001]: for shorter variable names, I would use "pseudolabel_" everywhere in place of "pseudo_label_"

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 4 above)
Comment (jwmueller) [reply tabular/src/autogluon/tabular/predictor/predictor.py:2465]: Why can't this method simply call `self._validate_fit_data()` somewhere inside of it, instead of duplicating the code from that other method?

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 6]
```diff
@@ -43,6 +43,7 @@
 # Done for Tabular
 # TODO: Remove all `time_limits` in project, replace with `time_limit`
 
+PSEUDO_MODEL_SUFFIX = "PSEUDO_{iter}"
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/predictor/predictor.py:46]: add comment to explain this like we have for DSTL suffix

File: core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py [hunk 7]
```diff
@@ -105,10 +108,26 @@ def _fit(self, model_base, time_start_fold, time_limit_fold, fold_ctx, kwargs):
...
-        fold_model.fit(X=X_fold, y=y_fold, X_val=X_val_fold, y_val=y_val_fold, time_limit=time_limit_fold, **kwargs_fold)
+
+            if is_pseudo:
+                # TODO: Add support for sample_weight when pseudo is present
+                logger.warning('WARNING: Sample weights given, but not used due to pseudo labelled data being given.')
+            else:
+                kwargs_fold['sample_weight'] = self.sample_weight[train_index]
+                kwargs_fold['sample_weight_val'] = self.sample_weight[val_index]
+
+        if is_pseudo:
```
Comment (DolanTheMFWizard) [inline core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py:123]: adds pseudo into train data

File: core/src/autogluon/core/trainer/abstract_trainer.py [hunk 8]
```diff
@@ -957,7 +964,14 @@ def _train_and_save(self, X, y, model: AbstractModel, X_val=None, y_val=None, st
                     time_left_total = time_limit
                 fit_log_message += f' Training model for up to {round(time_limit, 2)}s of the {round(time_left_total, 2)}s of remaining time.'
             logger.log(20, fit_log_message)
-            model = self._train_single(X, y, model, X_val, y_val, **model_fit_kwargs)
+
+            if X_pseudo is not None and y_pseudo is not None and X_pseudo.columns.equals(X.columns):
+                X_w_pseudo = pd.concat([X, X_pseudo])
```
Comment (DolanTheMFWizard) [inline core/src/autogluon/core/trainer/abstract_trainer.py:969]: Adds pseudo into train data for single models

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 1 above)
Comment (DolanTheMFWizard) [reply tabular/src/autogluon/tabular/predictor/predictor.py:857]: at lower levels like the SKLearn models level they take X and y separately. `train_data` gets split into X and y down the pipeline. I'm just doing it earlier.

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 2 above)
Comment (DolanTheMFWizard) [reply tabular/src/autogluon/tabular/predictor/predictor.py:978]: yea I couldn't think of a good name that covers everything. The name should ideally cover already pseudo labeled data coming in and non-labeled data

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 5 above)
Comment (DolanTheMFWizard) [reply tabular/src/autogluon/tabular/predictor/predictor.py:1001]: `pseudo_label_fit` is ideally what we want them to call because it runs the validation. `run_pseudolabeling` should be private, because there's no data validation or safety procedures in place.

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 4 above)
Comment (DolanTheMFWizard) [reply tabular/src/autogluon/tabular/predictor/predictor.py:2465]: so the issue is that we don't know if incoming data is labeled or not. That determines where it would go in `self._validate_fit_data()` if it is labeled, it should go into the `tuning_data` parameter, if not it should go into the `unlabled_data` parameter in the `_validate_fit_data()` signature. If the incoming data isn't in pandas format we can't check if the data is labeled or not, `_validate_pseudo_data` does this and handles both cases. There's a little redundancy, but the main thing is checking if data is labeled or not and checking it accordingly if data is not labeled and you pass it into `_validate_fit_data()` `tuning_data` param it throws and error and same with labeled data into the `unlabeled_data` parameter

**Comment by szha** [comment]
Job PR-1322-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/3/index.html

**Comment by szha** [comment]
Job PR-1322-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/4/index.html

**Comment by szha** [comment]
Job PR-1322-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/5/index.html

**Comment by szha** [comment]
Job PR-1322-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/6/index.html

File: tabular/src/autogluon/tabular/predictor/predictor.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/predictor/predictor.py:858]: The same argument could be said for train_data. Think about some reasons why we would want both to be contained in a single DataFrame. Think from the perspective of a novice ML / pandas user.

Comment (Innixma) [reply tabular/src/autogluon/tabular/predictor/predictor.py:858]: The one reason I could see us wanting y_pseudo to be separate is if we were doing soft labels for multiclass, but even then we could represent it via a single argument and passing in a tuple.

Comment (DolanTheMFWizard) [reply tabular/src/autogluon/tabular/predictor/predictor.py:858]: but for `fit_extra()` `train_data` is already split into X and y when retrieved by `self_trainer.load_data()`. So I was following that style.

**Comment by szha** [comment]
Job PR-1322-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/7/index.html

**Comment by szha** [comment]
Job PR-1322-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/8/index.html

**Comment by szha** [comment]
Job PR-1322-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/9/index.html

**Comment by szha** [comment]
Job PR-1322-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/10/index.html

**Comment by szha** [comment]
Job PR-1322-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/11/index.html

**Comment by szha** [comment]
Job PR-1322-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/12/index.html

**Comment by szha** [comment]
Job PR-1322-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1322/13/index.html
